### PR TITLE
Create an oEmbed endpoint for maps

### LINKED ISSUE #162: OEmbed support
Originally reported by: **Yohan Boniface (BitBucket: [yohanboniface](http://bitbucket.org/yohanboniface), GitHub: [yohanboniface](http://github.com/yohanboniface))**

---

http://www.oembed.com/

---
- Bitbucket: https://bitbucket.org/yohanboniface/umap/issue/162

### PR DESCRIPTION
**Pull request by davidbgk**
Fix #162

Based on https://oembed.com/

### CODE REVIEW

File: umap/views.py [hunk 1]
```diff
@@ -607,6 +619,52 @@ def render_to_response(self, context, *args, **kwargs):
...
+        if map_.share_status != Map.PUBLIC:
+            raise PermissionDenied("This map is not public.")
+
+        map_url = map_.get_absolute_url()
+        label = _("See full screen")
+        height = 300
+        data["height"] = height
+        width = 800
+        data["width"] = width
+        # TODISCUSS: do we keep width=100% by default for the iframe?
```
Comment (yohanboniface) [inline umap/views.py:657]: Can't we keep both width in relative ?

**Review by yohanboniface — APPROVED** [review]
Excellent!
Shame on me that I did not implement this earlier!

File: umap/tests/test_map_views.py [hunk 2]
```diff
@@ -675,3 +675,63 @@ def test_download_my_map(client, map, datalayer):
...
+
+
+def test_oembed_map(client, map, datalayer):
+    url = f"{reverse('map_oembed')}?url=http://testserver{map.get_absolute_url()}"
+    response = client.get(url)
+    assert response.status_code == 200
+    j = json.loads(response.content.decode())
+    assert j["type"] == "rich"
+    assert j["version"] == "1.0"
+    assert j["width"] == 800
```
[the remark this answers is not in the document]
Comment (davidbgk) [reply umap/tests/test_map_views.py:717]: The [spec](https://oembed.com/) says:

> The width/height in pixels required to display the HTML.

And maybe it was before we can have more flexible dimensions?

File: umap/views.py [hunk 3]
```diff
@@ -607,6 +619,52 @@ def render_to_response(self, context, *args, **kwargs):
         return response
 
 
+class MapOEmbed(View):
+    def get(self, request, *args, **kwargs):
+        data = {"type": "rich", "version": "1.0"}
+        format_ = request.GET.get("format", "json")
+        if format_ != "json":
+            response = HttpResponseServerError("Only `json` format is implemented.")
+            response.status_code = 501
```
[the remark this answers is not in the document]
Comment (davidbgk) [reply umap/views.py:628]: I tried to stick to the spec as much as possible:

> 501 Not Implemented
> The provider cannot return a response in the requested format. This should be sent when (for example) the request includes `format=xml` and the provider doesn't support XML responses. However, providers are encouraged to support both JSON and XML.

**Comment by bristow** [comment]
For the width, i give you 3 exemples : PeerTube, Youtube and osm.org

```
## PeerTube

<div data-oembed-url="https://tube-sciences-technologies.apps.education.fr/w/mhdTmfC5fgVq6WMjc5jp1g?start=10s"> 
 <div style="left:0;width:100%;height:0;position:relative;padding-bottom:56.25%"> 
  <iframe allowfullscreen="" src="https://tube-sciences-technologies.apps.education.fr/videos/embed/a43a9d48-75c0-4176-b336-3fd2b56e203b" style="top:0;left:0;width:100%;height:100%;position:absolute;border:0" tabindex="-1"></iframe> 
 </div> 
 
 ##Youtube
 <div data-oembed-url="https://www.youtube.com/watch?v=CHQrIMo0xak">
<div style="left: 0; width: 100%; height: 0; position: relative; padding-bottom: 56.25%;"><iframe allow="accelerometer; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share;" allowfullscreen="" scrolling="no" src="https://www.youtube.com/embed/CHQrIMo0xak?rel=0" style="top: 0; left: 0; width: 100%; height: 100%; position: absolute; border: 0;" tabindex="-1"></iframe></div>

## Openstreetmap
<div data-oembed-url="https://www.openstreetmap.org/#map=18/45.80318/4.83529">
<div style="left: 0; width: 100%; height: 0; position: relative; padding-bottom: 75%;"><iframe allowfullscreen="" src="https://www.openstreetmap.org/export/embed.html?bbox=4.832267761230469,45.80481138832744,4.835700988769531,45.80301636960864&amp;layer=mapnik" style="top: 0; left: 0; width: 100%; height: 100%; position: absolute; border: 0;" tabindex="-1"></iframe></div>
</div>
```

So 100% seems like a good choice

**Comment by davidbgk** [comment]
@bristow thank you so much for these example!

Do you also have the raw responses from the oembed endpoints of these services to see how they deal with the `width` parameter? The spec states that it should be in pixels so I'm curious how they handled that :)

**Comment by bristow** [comment]
Here is the response from the PeerTube developer

The oembed payload is generated here: https://github.com/Chocobozzz/PeerTube/blob/develop/server/core/controllers/services.ts#L99

For example: https://peertube2.cpy.re/services/oembed?url=https%3A%2F%2Fpeertube2.cpy.re%2Fw%2F1zywKcr1ChzL7R9rG6yCnq

Another example where we limit the size: https://peertube2.cpy.re/services/oembed?url=https%3A%2F%2Fpeertube2.cpy.re%2Fw%2F1zywKcr1ChzL7R9rG6yCnq&maxwidth=200

I hope this helps

**Comment by davidbgk** [comment]
> I hope this helps

Definitely!

I'm curious though because if you take the payload from https://peertube2.cpy.re/services/oembed?url=https%3A%2F%2Fpeertube2.cpy.re%2Fw%2F1zywKcr1ChzL7R9rG6yCnq
you have that value in the `html` key of the returned JSON:

```
<iframe width="560" height="315" 
  sandbox="allow-same-origin allow-scripts allow-popups" title="The Internet&#39;s Own Boy" 
  src="https://peertube2.cpy.re/videos/embed/04af977f-4201-4697-be67-a8d8cae6fa7a" 
  frameborder="0" allowfullscreen>
</iframe>
```

Which does not states `width="100%"` as [previously mentioned](https://github.com/umap-project/umap/pull/1526#issuecomment-1895350381). So where did you get previous HTML extracts?

**Comment by bristow** [comment]
I took these codes from the source code of [this site](https://martiniere-diderot.ent.auvergnerhonealpes.fr/), which offers remote media integration via oEmbed.

**Comment by bristow** [comment]
Will this feature be deployed in the near future? 

I'm willing to test it on a pre-prod platform if need be.

Thank you!

**Comment by yohanboniface** [comment]
Very soon!

We are in Javascript refactor which prevent for now to add more feature to the deployment flow, but it seems we are close the getting back to normal life! Sorry for that :)

**Comment by davidbgk** [comment]
@bristow this is now deployed on the OSM instance 🤞

**Comment by bristow** [comment]
Thank you so much!

But it doesn't seem to be working :/

Here's a capture of a render where I embed an umap link and below it an osm.org map link using the same method.

Sélection_1390

Any idea?

**Comment by bristow** [comment]
I analysed the source code of an umap map.
You can see the oembed URL starting with: https://umap.openstreetmap.fr/map/oembed/
But this URL returns a 404 error, whereas on a Peertube instance you get a json return: https://tube.ac-lyon.fr/services/oembed with the message that it can't find the URL


Perhaps a clue...

**Pull request merged**